doc/compiler/runtime/Recompilation.md

```diff
@@ -0,0 +1,142 @@
+in time" is to *re*compile methods "just in time". There are several 
+circumstances in when a method might be recompiled; these are outlined 
+in this doc.
+
+# Reasons for recompilation
+
+This section outlines the various reasons for the
+recompilation of a method. Recompilation can either be optional or 
+mandatory. However, since since any compilation is technically not
+mandatory for functional correctness (it exists for performance
+only), both *optional* and *mandatory* in the previous sentence should
+be suffixed with "in order to continue executing the existing compiled
+version of the method".
+
+## Optional Recompilation
+
+The following trigger a recompilation that is optional in the sense that
+it is not functionally incorrect to continue executing the existing JIT'd 
+body but there is a performance benefit to executing a more optimized body. 
+
+### Sampling
+
+The JIT has a Sampler Thread that is used to periodically perform tasks. One
+of these tasks is to set a flag on all Java Threads to yield at the next 
+yield point. Once the threads enter the yield point, they temporarily stop
+executing java code, and instead execute a JIT callback (see 
+`jitMethodSampleInterrupt` in `HookedByTheJit.cpp`). This callback is used
+to build a window of samples, which is used to determine whether or not a 
+method should be recompiled.
+
+### Counting
+
+This mechanism refers to the notion of triggering a recompilation by counting
+the number of times the method is invoked. Guarded Counting Recompilation (GCR)
+is an example of this, wherein the compiler will generate Trees at the start of
+of a method to decrement a counter when the method is invoked, and to trigger
+a recompilation when the counter hits 0. A "Counting Body" is another example
+which is similar in principle to GCR, but has a different pre-prologue. 
+
+### Explict
+
+This section refers to mechanisms that trigger recompilation explicitly, i.e.
+not based on counting or sampling, but via some other criteria. The features
+in OpenJ9 that might explicitly trigger recompilation are:
+* Exception Driven Optimization (EDO)
+* JProfiling
+* On Stack Replacement (OSR)
+* Hardware Profiling
+
+To use OSR as an example, the OSR mechanism may force execution to transition
+to the interpreter because an optimized path in the generated code is no longer
+valid to run. As part of this transition, the mechanism might also trigger a
+recompilation if this path contributed to a significant porition of the 
+performance provided by the method - i.e. it is better to recompile the
+method than leave it as is.
+
+## Mandatory Recompilation
+
+The following trigger a recompilation that is mandatory in the sense
+that 1. it **is** functionally incorrect to execute the existing JIT'd body,
+and 2. in order to (at the very least) maintain the performance provided by 
+the previously valid compiled code.
+
+### Invalidation 
+
+Invalidation refers to circumstance when assumptions made about the
```

Comment:
Pre-existence is a major example for a mechanism that causes bodies to become invalid, so it deserves a mention.